lib/router/src/cache.rs

```diff
@@ -1,426 +1,461 @@
+struct Node<T> {
+    dq_key: delay_queue::Key,
+    value: T,
 }
 
 // ===== impl Cache =====
 
-impl<K: Hash + Eq, V> Cache<K, V, ()> {
-    pub fn new(capacity: usize, max_idle_age: Duration) -> Self {
-        Self {
+impl<K, V> Cache<K, V>
+where
+    K: Clone + Eq + Hash,
+    V: Clone,
+{
+    pub fn new(capacity: usize, expires: Duration) -> (Lock<Cache<K, V>>, PurgeCache<K, V>) {
+        assert!(capacity != 0);
+        let cache = Self {
             capacity,
-            vals: IndexMap::default(),
-            max_idle_age,
-            now: (),
-        }
-    }
-}
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
+            values: IndexMap::default(),
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
 
-impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
-    /// Accesses a route.
-    ///
-    /// A mutable reference to the route is wrapped in the returned `Access` to
-    /// ensure that the access-time is updated when the reference is released.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
-    where
-        Q: Hash + Equivalent<K>,
-    {
-        let v = self.vals.get_mut(key)?;
-        Some(v.access(&self.now))
+        (cache, bg_purge)
     }
 
-    /// Ensures that there is capacity to store an additional route.
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
-        if self.vals.len() == self.capacity {
-            // Only whole seconds are used to determine whether a node should be retained.
-            // This is intended to prevent the need for repetitive reservations when
-            // entries are clustered in tight time ranges.
-            let max_age = self.max_idle_age.as_secs();
-            let now = self.now.now();
-            self.vals.retain(|_, n| {
-                let age = now - n.last_access();
-                age.as_secs() <= max_age
-            });
-
-            if self.vals.len() == self.capacity {
-                return Err(CapacityExhausted {
-                    capacity: self.capacity,
-                });
-            }
-        }
-
-        Ok(Reserve {
-            vals: &mut self.vals,
-            now: &self.now,
-        })
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
 
-    /// Overrides the time source for tests.
-    #[cfg(test)]
-    fn with_clock<M: Now>(self, now: M) -> Cache<K, V, M> {
-        Cache {
-            now,
-            vals: self.vals,
-            capacity: self.capacity,
-            max_idle_age: self.max_idle_age,
+    pub fn get(&mut self, key: &K) -> Option<V> {
```

Comment:
tioli: I prefer `access` to get here, since it's `&mut self` --- `get`s are typically `&self`; we actually mutate state, so it would be clearest to use a verb that makes that clearer.

Also, would be nice to document the semantics of this function.... Something like:

```
/// Attempts to access an item by key.
///
/// If a value is returned, this key will not be considered for eviction for... (how long?)
```